lib/smack/DSAWrapper.cpp

```diff
@@ -214,24 +102,104 @@ unsigned DSAWrapper::getPointedTypeSize(const Value *v) {
     llvm_unreachable("Type should be pointer.");
 }
 
-int DSAWrapper::getOffset(const Value *v) {
-  return getOffset(new MemoryLocation(v));
+unsigned DSAWrapper::getOffset(const Value *v) {
+  if (!DG->hasCell(*v))
+    return 0;
+  return DG->getCell(*v).getOffset();
+}
+
+const sea_dsa::Node *DSAWrapper::getNode(const Value *v) {
+  // For sea-dsa, a node is obtained by getting the cell first.
+  // It's possible that a value doesn't have a cell, e.g., undef.
+  if (!DG->hasCell(*v))
+    return nullptr;
+  auto node = DG->getCell(*v).getNode();
+  assert(node && "Values should have nodes if they have cells.");
+  return node;
 }
 
-// TODO: Should this return the node or its leader?
-const DSNode *DSAWrapper::getNode(const Value *v) {
-  const llvm::EquivalenceClasses<const llvm::DSNode *> &eqs =
-      nodeEqs->getEquivalenceClasses();
-  auto *N = nodeEqs->getMemberForValue(v);
-  return N ? eqs.getLeaderValue(N) : nullptr;
+bool DSAWrapper::isTypeSafe(const Value *v) {
```

Comment:
Why is the argument of this function of type `Value` and not of type `Node`?
Now that I look at this code, I don't understand why it should be `Value` because `Node` would seem more appropriate - since this function just uses `v` to get its node and for nothing else.

Reply:
This function actually further get the offset. Type safety is a property with respect to offsets, which means that just getting the node is not enough.